exoscale/client.go

```diff
@@ -1,29 +1,162 @@
 package exoscale
 
 import (
+	"context"
+	"encoding/json"
 	"errors"
 	"fmt"
 	"os"
+	"path/filepath"
+	"sync"
 
 	"github.com/exoscale/egoscale"
+	"gopkg.in/fsnotify.v1"
 )
 
 const defaultComputeEndpoint = "https://api.exoscale.com/v1"
 
-func newExoscaleClient() (*egoscale.Client, error) {
+type exoscaleClient struct {
+	client          *egoscale.Client
+	credentialsFile string
+	endpoint        string
+	*sync.RWMutex
+}
+
+type exoscaleAPICredentials struct {
+	APIKey    string `json:"api_key"`
+	APISecret string `json:"api_secret"`
+	Name      string `json:"name"`
+}
+
+func newExoscaleClient(stop <-chan struct{}) (*exoscaleClient, error) {
+	exoclient := &exoscaleClient{RWMutex: &sync.RWMutex{}, endpoint: defaultComputeEndpoint}
+
 	envEndpoint := os.Getenv("EXOSCALE_API_ENDPOINT")
 	envKey := os.Getenv("EXOSCALE_API_KEY")
 	envSecret := os.Getenv("EXOSCALE_API_SECRET")
+	envCredentialsFile := os.Getenv("EXOSCALE_API_CREDENTIALS_FILE")
 
-	if envEndpoint == "" {
-		envEndpoint = defaultComputeEndpoint
+	if envEndpoint != "" {
+		exoclient.endpoint = envEndpoint
 	}
 
-	if envKey == "" || envSecret == "" {
+	egoscale.UserAgent = fmt.Sprintf("Exoscale-K8s-Cloud-Controller/%s %s", versionString, egoscale.UserAgent)
+
+	if envKey != "" && envSecret != "" {
+		infof("reading exoscale IAM credentials from environment")
+		exoclient.client = egoscale.NewClient(exoclient.endpoint, envKey, envSecret)
+	} else if envCredentialsFile != "" {
+		exoclient.credentialsFile = envCredentialsFile
+		infof("reading exoscale IAM credentials from file %q", exoclient.credentialsFile)
+		exoclient.refreshCredentials()
+		go exoclient.watchCredentialsFile(stop)
+	} else {
 		return nil, errors.New("incomplete or missing for API credentials")
```

Comment:
Not changed in your PR, but it's a good opportunity to fix this typo :)

```suggestion
		return nil, errors.New("incomplete or missing Exoscale API credentials")
```